### ISSUE TITLE
check that MCS invariants imply orphanage invariants

### ISSUE THREAD

**Issue by lsf37** [seL4/l4v#456]
The "Orphanage" theory in the Refine session was added to make sure that thread states correspond to the scheduler queues, i.e. that no running threads can become orphaned by being left out of the scheduler queues.

AFAIR, the standard MCS invariants already include this property (and on the abstract spec level already, because these queues are now visible there), but I'm not sure we state this in a form that is obviously equivalent with `no_orphans`.

For reference `no_orphans` is

```isabelle
 ∀ tcb_ptr.
         tcb_ptr : all_active_tcb_ptrs s
         ⟶
         tcb_ptr = ksCurThread s ∨ tcb_ptr : all_queued_tcb_ptrs s ∨
         ksSchedulerAction s = SwitchToThread tcb_ptr
```

This issue is for checking what the actual status is and potentially adding a lemma to the abstract invariants in MCS that show something like `invs s ==> no_orphans s`

**Comment by lsf37** [comment]
A bit more info on this one: the property as stated does not hold on MCS, because there are more (intended) cases. What I think does hold (which is pretty much directly `valid_blocked`, which is part of the abstract invariants): 

all runnable threads t *with an active SC* are either:
- the current thread
- in a scheduling queue
- in the release queue
- or about to be switched to (sched action = SwitchToThread t)

I think that is as close to a no-orphan property as MCS can get.

It would be nice to know that once you are in the release queue you will eventually end up in a scheduling queue, but that is only true if nothing happens to the thread in the meantime (deletion, removed SC, blocked, etc).

So unless somebody disagrees, I think we can close this and delete the orphanage on the `rt` branch.

**Comment by Xaphiosis** [comment]
I have read both the original issue and your message, but I want to make sure I understand:

The original says that the lack of orphans might be implied by the existing MCS invariants.
The follow-up says that there are more cases, but that on MCS `valid_blocked` implies a number of interesting properties for threads with an SC.
The idea of no orphans was (I think) that no active thread is lost.

Is the missing link that no thread without an SC can be considered active? If so, then we can indeed clear out the orphanage (sounds macabre) in the MCS proof.

The remaining question from my side is whether you still want me to complete the missing arches of orphanage on master.

**Comment by lsf37** [comment]
> The original says that the lack of orphans might be implied by the existing MCS invariants. The follow-up says that there are more cases, but that on MCS `valid_blocked` implies a number of interesting properties for threads with an SC. The idea of no orphans was (I think) that no active thread is lost.
>
> Is the missing link that no thread without an SC can be considered active? 

Yes, that is basically it (it also needs to be an active SC, not just any SC). Overall, the question of what an active thread is becomes more interesting in MCS, so the property is no longer a nice and clear as before.

There is an additional wrinkle that what the kernel does with active threads is also more complicated: the same cases as the no_orphans property, plus the thread can also be in the release queue. But that is minor, the main difference is the SC.

> If so, then we can indeed clear out the orphanage (sounds macabre) in the MCS proof.

I think so, yes.

> The remaining question from my side is whether you still want me to complete the missing arches of orphanage on master.

To me that is still worth doing if you can stomach it. It'll make proofs on master less painful until MCS completely takes over, which may still be many years in the future.
